### DISCUSSION TITLE
State of test automation?

### DISCUSSION THREAD

**Discussion by mfenniak** [pgsql-io/multicorn2#44] [General]
Is it expected in the current `main` branch that the tests work?  eg. `make installcheck`, or the `docheck` script?  I haven't been able to get them to run successfully, but, I admit to using NixOS where everything is a little quirky.

Would there be any interest in merging a patch that would...
- Provide a way to run all the tests, across a matrix of different environments (python versions) * (postgresql versions)
- Run the tests automatically in PRs using GitHub Actions
- And do it by using the Nix package manager

The side-effects of the last point should be spelt out a little more.  It would mean that a developer could run all the tests using this configuration only if they installed the Nix package manager (available for any Linux distro, MacOS, or NixOS).  The environments available would be limited to those supported by Nix (eg. PostgreSQL 12-16 currently, Python 3.8 - 3.13 currently).  And, if no other developers are familiar with Nix, it becomes difficult to improve or maintain these changes.

Nix is a good tool for this type of usage because it allows for a combination of a dynamic build and testing environment across all the different versions of those dependencies without having to install them in a way that makes them available system-wide.  It's certainly not the only good approach, but it's one that I'm comfortable with and find powerful.

As an example of some part of what this would look like, I've packaged multicorn2 with nix so that I could use it to create a new docker image for dynamodb_fdw.  The multicorn2 packaging is https://github.com/mfenniak/custom-nixpkgs/blob/main/multicorn2.nix; so some code similar to this, although configured to run `make installcheck` across all these different environments, would be what I'd be proposing to add.

I'd enjoy the work in creating such a patch, but I thought it best to discuss such a thing before doing the work.  If it wasn't a welcome contribution, I wouldn't want to blow through my time. :-)

**Comment by ShaheedHaque** [comment]
I am happy to see increased focus on the tests etc. (tbh, I had not even noticed the test Make targets you mentioned), but for clarity, I defer to @luss on this.

If taken forward, I'd ideally like:

1. The tests to always run on a "make". 
2. The new nix-based way of running tests to replace the old way(s)... Or rehost the old ways on nix.

**Comment by luss** [comment]
we DEFINTELY need and want more testing.  Else it's impractical to fix
minor issues because of the delicacy of multiplatform python and the
short-sheet effect

**Comment by luss** [comment]
Lets just support current versions of Python 3.9++, and current versions of Postgres 12++.   I've opened up v3.0devel with these base changes and now (I feel) we are in poisition to modernize our tooling and to fix our bugs (like predicate pushdown)

**Comment by mfenniak** [comment]
Thanks for sharing your thoughts.

I think #46 misses on point 1 in @ShaheedHaque's comment.  It doesn't quite make it so that "tests always run on a make" -- although it might be possible to change the default target to support that -- but it does document the test capability in the README, and bring test failures to visibility during merge.  You could just use the nix test run as a build command... depending on what you're doing after the build. 😉

**Comment by luss** [comment]
I agree (as Mathieu suggested) that Nix testing replaces the docker crap.   Mathieu please whack the docker testing stuff as per the commit u aborted.   If anyone needs it, we can always go back and create a REL2_STABLE branch.  We are moving forward smartly.  Great stuff!!

**Comment by luss** [comment]
Very kewl!!   Unless someone abjects.   I will tag a devel release and try
and do some tests build on pg16 and pg17.    pg17 comes out bneta 1 later
this week.

**Comment by luss** [comment]
In v3.1, just recently opened up for devel, it is my objective that we support pg17 & py3.12.  EZ for me to day...   I sure hope someone will run/write some tests that (at least mostly) prove it.  AFAIK, we've fixed any weirdown pushdown (or lack therof) erros for recent PG releases.  If not, if someone can isolate a iple use cases that doesn't work.  I'll get a PG internals hacker to look at our c code and figure out what's wrong.

**Reply by mfenniak** [reply]
Running the existing test suite automatically on new changes against more versions of Python & PostgreSQL will require very few modifications (assuming, of course, that they pass):
- `nix flake update` at the root allows access to the latest releases of PostgreSQL and Python that are packaged by nixpkgs
- The target test PG & PY versions are defined here: https://github.com/pgsql-io/multicorn2/blob/3a77c129aa0ae865e3cfebfcd577fef53ec9e8fb/flake.nix#L24-L37

**Comment by luss** [comment]
:-)   I will confess to not knowing how to run the existing tests.  Have we
documented it somewhere so I can try it?

**Reply by mfenniak** [reply]
You bet; it's noted in the README.  https://github.com/pgsql-io/multicorn2?tab=readme-ov-file#integration-tests

**Comment by luss** [comment]
ahh yesss!   I do rember reading about multi-version python nix teasting in
the README.  I'm old and forgetful.  :-)
